RavenDB-18852 Support range queries using the index instead of scan

ayende opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18852 

### Additional description

This PR adds support for running range queries by using `MultiTermProvider`.
We scan through the index and do efficient range queries to provide the relevant terms.

I provided only the base infrastructure, still need to wire it in all the right places. 

Also added support for storing `FixedSizeTree` where the key is a `double`. 

### Type of change

- New feature

### How risky is the change?

- Not relevant

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works


### Is there any existing behavior change of other features due to this change?

Need to wire that to the Corax queries execution

### UI work

- No UI work is needed

ayende committed:
'RavenDB-18852 - Properly building range term matches to support less than, between and greater than

ayende committed:
RavenDB-18852 - Adding support for doing indexed range queries on top of numeric data (int64)

ayende committed:
RavenDB-18852 - Moving FixedSizeTree to be generic (preparing to add FixedSizedTree<double>!)

ayende committed:
RavenDB-18852 - Can now support queries on numeric using doubles as well

ayende commented on bench/Micro.Benchmark/Micro.Benchmark.csproj:
```diff
@@ -4,6 +4,8 @@
     <RuntimeFrameworkVersion>6.0.6</RuntimeFrameworkVersion>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <AssemblyName>Micro.Benchmark</AssemblyName>
+    <EnablePreviewFeatures>True</EnablePreviewFeatures>
```
I'm using `Generic Math` here, so we have to use the preview feature.
@arek / @ppekrol - I don't _expect_ that to be an issue, but wanted your eyes on that as well.

ayende commented on src/Corax/IndexFieldsMapping.cs:
```diff
@@ -30,11 +30,37 @@ public IndexFieldsMapping(ByteStringContext context)
     }
 
+    
+    private const int I64Postfix = 875972909; //"-I64"
+    private const int DblPostfix = 1818379309; //"-Dbl"
```
This is a bit different from what we do for Lucene. Where the suffixes are `_L` and `_D`. I don't think it matters, but wanted to point it out.

ayende replied:
Changed that to match Lucene.

ayende commented on src/Corax/IndexFieldsMapping.cs:
```diff
@@ -30,11 +30,37 @@ public IndexFieldsMapping(ByteStringContext context)
         if (!_fieldsById.TryGetValue(fieldId, out var storedAnalyzer))
         {
-            var binding = new IndexFieldBinding(fieldId, fieldName, analyzer, hasSuggestion, fieldIndexingMode, hasSpatial);
+            GetFieldNameForDoubles(fieldName, out var fieldNameDouble);
+            GetFieldNameForLongs(fieldName, out var fieldNameLong);
```
@maciejaszyk Here we _have_ to check that the lifetime of the `IndexFeidlsMapping` is as long as the `context`.
I'm fairly certain that it is, but I want another pair of eyes on that.

ayende commented on src/Corax/IndexSearcher/IndexSearcher.MultiTermMatches.cs:
```diff
@@ -321,6 +366,96 @@ public AndNotMatch NotInQuery<TInner>(string field, TInner inner, List<string> n
+            (false, NullScoreFunction) => MultiTermMatch.Create(new MultiTermMatch<TermRangeProvider<TLow, THigh>>(_transaction.Allocator,
+                new TermRangeProvider<TLow, THigh>(this, terms,fieldName, low, high))),
+
+            _ => throw new NotSupportedException()
+            // (true, NullScoreFunction) => MultiTermMatch.Create(new MultiTermMatch<NotStartWithTermProvider>(_transaction.Allocator,
+            //     new NotStartWithTermProvider(this, _transaction.Allocator, terms, fieldName, fieldId, slicedTerm))),
+            //
+            // (false, _) => MultiTermMatch.Create(
+            //     MultiTermBoostingMatch<StartWithTermProvider>.Create(
+            //         this, new StartWithTermProvider(this, _transaction.Allocator, terms, fieldName, fieldId, slicedTerm), scoreFunction)),
+            //
+            // (true, _) => MultiTermMatch.Create(
+            //     MultiTermBoostingMatch<NotStartWithTermProvider>.Create(
+            //         this, new NotStartWithTermProvider(this, _transaction.Allocator, terms, fieldName, fieldId, slicedTerm), scoreFunction))
```
Need to fill this in.

ayende commented on src/Corax/IndexSearcher/IndexSearcher.MultiTermMatches.cs:
```diff
@@ -321,6 +366,96 @@ public AndNotMatch NotInQuery<TInner>(string field, TInner inner, List<string> n
+            (false, NullScoreFunction) => MultiTermMatch.Create(new MultiTermMatch<TermNumericRangeProvider<TLow, THigh, long>>(_transaction.Allocator,
+                new TermNumericRangeProvider<TLow, THigh, long>(this, set, terms, fieldName, low, high))),
+
+            _ => throw new NotSupportedException()
```
Need to fill this in.

ayende commented on src/Corax/IndexSearcher/IndexSearcher.MultiTermMatches.cs:
```diff
@@ -321,6 +366,96 @@ public AndNotMatch NotInQuery<TInner>(string field, TInner inner, List<string> n
+            (false, NullScoreFunction) => MultiTermMatch.Create(new MultiTermMatch<TermNumericRangeProvider<TLow, THigh, double>>(_transaction.Allocator,
+                new TermNumericRangeProvider<TLow, THigh, double>(this, set, terms, fieldName, low, high))),
+
+            _ => throw new NotSupportedException()
```
Need to fill this in.

ayende commented on src/Corax/IndexWriter.cs:
```diff
@@ -84,8 +87,14 @@ private IndexWriter(IndexFieldsMapping fieldsMapping)
             for (int i = 0; i < bufferSize; ++i)
+            {
                 _buffer[i] = new Dictionary<Slice, List<long>>(SliceComparer.Instance);
+                _bufferDoubles[i] = new Dictionary<double, List<long>>();
+                _bufferLongs[i] = new Dictionary<long, List<long>>();
+            }
```
we should probably add a cache here for those, no?

redknightlois replied:
Yes, but we havent applied allocations optimizations yet. We only applied for the most blatant offenders. There is lots of opportunities to improve that, and this is just one of those. I wouldnt worry much for now.

ayende commented on src/Corax/Queries/TermProviders/TermProvider.Range.cs:
```diff
@@ -0,0 +1,241 @@
+        private readonly FixedSizeTree<TVal>.IFixedSizeIterator _iterator;
+        private bool _first;
+
+        private const int TermBufferSize = 32;
```
`long.MaxValue` len in chars == 19
`double.MaxValue` len in chars == 21

ayende commented on src/Corax/Queries/UnaryMatch.Ordinal.cs:
```diff
@@ -947,7 +947,7 @@ private static int FillFuncNumericalAll<TComparer>(ref UnaryMatch<TInner, TValue
                     in inner, UnaryMatchOperation.GreaterThan,
                     searcher, fieldId, value,
-                    mode == UnaryMatchOperationMode.Any ? &FillFuncNumericalAny<GreaterThanMatchComparer> : &FillFuncNumericalAny<GreaterThanMatchComparer>, &AndWith,
+                    mode == UnaryMatchOperationMode.Any ? &FillFuncNumericalAny<GreaterThanMatchComparer> : &FillFuncSequenceAll<GreaterThanMatchComparer>, &AndWith,
```
Pretty sure that this is a valid fix, but I want another pair of eyes here.

maciejaszyk replied:
This is the builder for numerics. Notice the above condition when we check if `TValueType` is a `Slice`.

ayende replied:
So this should be `FillFuncNumericalAll` ?

maciejaszyk replied:
I think so

redknightlois replied:
Yes.

ayende commented on src/Corax/Queries/TermProviders/TermProvider.Range.cs:
```diff
@@ -0,0 +1,241 @@
+    public unsafe struct TermNumericRangeProvider<TLow, THigh, TVal> : ITermProvider
+        where TLow : struct, Range.Marker
+        where THigh  : struct, Range.Marker
+        where TVal : unmanaged, IBinaryNumber<TVal>, IMinMaxValue<TVal>, INumber<TVal>
```
This is where we use generic math

ayende committed:
RavenDB-18852 - Match numeric suffixes with their Lucene versions

redknightlois commented on src/Corax/IndexWriter.cs:
```diff
@@ -244,10 +254,11 @@ void AddMaybeAvoidDuplicate(List<long> term, long entryId)
 
                     ExactInsert(valueInEntry);
+                    NumericInsert(lVal, dVal);
                     break;
```
Try `NumericInsert` on top instead, looks like something with the parser.

ayende replied:
I don't follow? 
This insert a numeric token

ayende committed:
RavenDB-18852 - Fixing wrong function being used

redknightlois commented on src/Corax/IndexWriter.cs:
```diff
@@ -677,25 +649,144 @@ public void Commit()
+        {
+            FixedSizeTree fieldTree = fieldsTree.FixedTreeFor(_fieldsMapping.GetByFieldId(fieldId).FieldNameLong, sizeof(long));
+            var llt = Transaction.LowLevelTransaction;
+            var sortedTerms = _bufferLongs[fieldId].Keys.ToArray();// TODO: let's avoid this allocation
```
We haven't done allocations optimizations for the most part. Dont worry about it.

maciejaszyk committed:
RavenDB-18820 Fixing deletions in Corax (partial)

ayende committed:
RavenDB-18852 - Fixing compilation - moving local function to its right location

ayende committed:
RavenDB-18852
- Ensuring that after defrag of a container page, we take into account that the required size may be different
- Moved debug only checks to runtime checks, to avoid data corruptions

redknightlois committed:
RavenDB-18875: Buffer passed is smaller than actual size.

We were trying to be too clever and recorded sizes & actual buffer sizes may desync. The actual buffer size is a much better indication if the buffer will be able to hold the value or not.

redknightlois committed:
RavenDB-18690: IndexEntryWriter is not returning the buffer to the pool

redknightlois committed:
RavenDB-18820 Avoid passing around reference to internal buffers

redknightlois committed:
RavenDB-18852: Updating size before releasing the temporary arrays.

redknightlois committed:
RavenDB-18852 - Trying to reduce buffer management in indexing

redknightlois committed:
RavenDB-18820: Merging deletions work with numeric support.

ayende committed:
RavenDB-18852 Fixing the way we are handling internal free list inside of the Container (previously we would keep a reference to a page after we freed it)

redknightlois commented:
Can be closed now. Will be merged from https://github.com/ravendb/ravendb/pull/14492
